lang/codegen/src/generator/dispatch.rs

```diff
@@ -0,0 +1,697 @@
+        quote! {
+            [ #( #selector_bytes ),* ] => {
+                Ok(Self::#variant_ident(
+                    #(
+                        <#variant_types as ::scale::Decode>::decode(input)?
+                    ),*
+                ))
+            }
+        }
+    }
+
+    /// Generates one match arm of the dispatch variant enum.
+    ///
+    /// # Note
+    ///
+    /// There is one match arm per ink! constructor or message for the dispatch
+    /// `scale::Decode` implementation.
+    fn generate_dispatch_variant_arm<C>(
+        &self,
+        cws: ir::CallableWithSelector<'_, C>,
+    ) -> TokenStream2
+    where
+        C: ir::Callable,
+    {
+        let input_types = cws.callable().inputs().map(|arg| &arg.ty);
+        let variant_ident = self.generate_dispatch_variant_ident(cws);
+        quote! {
+            #variant_ident(#(#input_types),*)
+        }
+    }
+
+    /// Returns `true` if all ink! messages of `self` deny payments.
+    ///
+    /// # Note
+    ///
+    /// This information is used to produce better code in this scenario.
+    fn all_messages_deny_payment(&self) -> bool {
+        self.contract
+            .module()
+            .impls()
+            .flat_map(ir::ItemImpl::iter_messages)
+            .all(|message| !message.is_payable())
+    }
+
+    /// Generates one match arm of the dispatch message for the `execute` implementation.
+    ///
+    /// # Note
+    ///
+    /// This is basically the code per ink! message that is going to be executed after
+    /// the dispatch has already taken place.
+    fn generate_dispatch_execute_message_arm(
+        &self,
+        cws: ir::CallableWithSelector<'_, ir::Message>,
+    ) -> TokenStream2 {
+        let storage_ident = self.contract.module().storage().ident();
+        let ident = self.generate_dispatch_variant_ident(cws);
+        let message = cws.callable();
+        let arg_pats = message.inputs().map(|arg| &arg.pat).collect::<Vec<_>>();
+        let arg_inputs = if arg_pats.len() == 1 {
+            quote! { #(#arg_pats),* }
+        } else {
+            quote! { ( #(#arg_pats),* ) }
+        };
+        let (mut_mod, msg_trait, exec_fn) = match message.receiver() {
+            ir::Receiver::RefMut => {
+                (
+                    Some(quote! { mut }),
+                    quote! { MessageMut },
+                    quote! { execute_message_mut },
+                )
+            }
+            ir::Receiver::Ref => {
+                (None, quote! { MessageRef }, quote! { execute_message })
+            }
+        };
+        let selector_id = cws.composed_selector().unique_id();
+        let namespace = Self::dispatch_trait_impl_namespace(ir::CallableKind::Message);
+        // If all ink! messages deny payment we can move the payment check to before
+        // the message dispatch which is more efficient.
+        let accepts_payments = cws.is_payable() || self.all_messages_deny_payment();
```

Comment:
```suggestion
        let accepts_payments = cws.is_payable();
```

Reply:
This change would be wrong. If all messages deny payment we can move the check for each of them individually to before dispatch which is more optimal in this case.

Reply:
My confusion stems from the naming of `accepts_payments`, it appears illogical. Now I understand what you mean, that if `self.all_messages_deny_payment() == true`, then the check happens here:

(https://github.com/paritytech/ink/blob/7f1bb35f1ed2652e5eb07e0cd5e6728a564909ea/lang/codegen/src/generator/dispatch.rs#L110) 

However, even with the comment this was not immediately obvious. Might be better to be named `check_accepts_payments` or similar.

Reply:
Yeah even I was confused during development. Maybe it really is the bad naming here.

Reply:
Maybe renaming to `deny_payments` is better here. However, normally I try to formulate everything "positively" with regards to checks etc. to avoid statements like "check against non unpayable non-message". :D

Reply:
I think that's just equivalent to `accept_payments`. The problem is that `true` either means "don't perform the check because this method is payable" or "don't perform the check because it is done somewhere else". I get that the comment is supposed to explain that, but I didn't get it at first.

Reply:
Yeah I can totally understand your confusion. I am confused as well. :S Not sure what name would be better though.